packages/graph/migrations/postgres/migration/1656417312397_initial.ts

```diff
@@ -376,89 +378,93 @@ export async function up(pgm: MigrationBuilder): Promise<void> {
     primaryKey: ["entity_id", "version"],
   });
 
+  pgm.addConstraint("entities", "entities_relation_constraint", {
+    check: `(left_entity_id IS NULL     AND right_entity_id IS NULL) 
+          OR left_entity_id IS NOT NULL AND right_entity_id IS NOT NULL`,
+  });
+
   pgm.createTable(
-    "entity_relations",
+    "entity_histories",
     {
-      source_entity_id: {
+      entity_id: {
         type: "UUID",
-        notNull: true,
         references: "entity_ids",
-      },
-      target_entity_id: {
-        type: "UUID",
         notNull: true,
-        references: "entity_ids",
       },
-      link_entity_id: {
-        type: "UUID",
+      version: {
+        type: "TIMESTAMP WITH TIME ZONE",
         notNull: true,
-        references: "entity_ids",
+        default: pgm.func("clock_timestamp()"),
       },
-      created_by_id: {
+      entity_type_version_id: {
         type: "UUID",
         notNull: true,
-        references: "accounts",
+        references: "entity_types",
       },
-      created_at: {
-        type: "TIMESTAMP WITH TIME ZONE",
+      properties: {
+        type: "JSONB",
         notNull: true,
-        default: pgm.func("clock_timestamp()"),
       },
-    },
-    {
-      ifNotExists: true,
-    },
-  );
-  // Currently entity relations are between unversioned entities.
-  pgm.addConstraint("entity_relations", "entity_relations_pkey", {
-    primaryKey: ["source_entity_id", "link_entity_id", "target_entity_id"],
-  });
-
-  pgm.createTable(
-    "entity_relation_histories",
-    {
-      // We should consider whether these should reference entity_ids or not.
-      // If we allow GDPR removal of entities, this constraint has to fail/cascade depending on desired output.
-      source_entity_id: {
+      left_order: {
+        // TODO: this is where we could do fractional indexing
+        //  https://app.asana.com/0/1200211978612931/1202085856561975/f
+        type: "integer",
+        notNull: false,
+      },
+      right_order: {
+        // TODO: this is where we could do fractional indexing
+        //  https://app.asana.com/0/1200211978612931/1202085856561975/f
+        type: "integer",
+        notNull: false,
+      },
+      left_entity_id: {
         type: "UUID",
-        notNull: true,
+        notNull: false,
         references: "entity_ids",
       },
-      target_entity_id: {
+      right_entity_id: {
         type: "UUID",
-        notNull: true,
+        notNull: false,
         references: "entity_ids",
       },
-      link_entity_id: {
+      archived: {
+        // We can consider allowing null values to reduce used space.
```

Comment:
Any resources on this?

Reply:
I presume we'd want to set `notNull: true` if we just stick with booleans

Reply:
I found it here https://dba.stackexchange.com/a/300789
I couldn't find the docs the original poster referred to. Perhaps it's a combination of a boolean taking up 1 byte and nulls taking up 1 bit in a bitmap.

There's also this concluding the same: https://www.postgresql.org/message-id/BANLkTimCGA9qDsAorBOr=FqibRVxhjhkHw@mail.gmail.com

Reply:
Sounds like we should use `NULL`s to me. I'm guessing there's no marker type we could use instead of boolean? (e.g. a `NONNULL` value)

Just slightly concerned that it'd be too easy to also put in a `FALSE value` and then our querying would be a bit buggered

Reply:
The extra annoyance is having to do the null check with `x AND x IS NOT NULL` rather than just `x` in all queries.
I am almost tempted to go with the `bit` data type instead so we gain the benefit of NULL size savings, but also having to only query by `x = 0` or `x = 1`.

Thoughts on that?

Reply:
The `bit` data type is what I meant. I assumed we were basically saying:
`NULL` or `TRUE` and just trying to avoid `FALSE`. (Which seems bad to me)

Reply:
I didn't notice i had set notNull to false here.. What a strange key. I updated to use a bit instead :)